keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,388 @@
+- [Performance degradation in large k8s deployments.][perf-degrade] Not being able to efficiently read/update individual endpoint changes can lead to (e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If one consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+### Goal
+
+- Support tens of thousands of backend endpoints in a single service on cluster with thousands of nodes.
+- Move the API towards a general-purpose backend discovery API.
+- Leave room for foreseeable extension:
+  - Support multiple IPs per pod
+  - More endpoint states than Ready/NotReady
+  - Dynamic endpoint subsetting
+  
+### Non-Goal
+- Change functionality provided by K8s V1 Service API.
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
+The following new EndpointSlice API will be added to the `Discovery` API group.
+
+```
+type EndpointSlice struct {
+    metav1.TypeMeta
+    // OwnerReferences should be set when the object is derived from a k8s service.
+    // The object labels may contain the following well known keys:
+    // kubernetes.io/service: the label value indicates the name of the service from which the EndpointSlice is derived from.
+    metav1.ObjectMeta
+    // Metadata of the Endpoints in the EndpointSlice
+    EndpointMeta
+    // Endpoints of this EndpointSlice
+    Endpoints []Endpoint
+}
+
+type Endpoint struct {
+    // Required: must contain at least one backend.
+    // The type of backend must conforms with the BackendType specified in SliceMeta.
+    // Backend can be an IP, URL or hostname.
+    // Different consumers (e.g. kube-proxy) handle different types of backends in the context of its own capabilities.
+    Backends []string
+    // Optional: the conditions of the endpoint
+    Condition EndpointCondition
+    // Optional: Reference to object providing the endpoint.
+    TargetRef *v1.ObjectReference
+    // Optional: this can contain arbitrary metadata associated with the endpoint.
+    // The key must conform with the label format.
+    // The labels may contain the following well known keys:
+    // kubernetes.io/node: the value indicates the corresponding node where the endpoint is located.
+    // kubernetes.io/pod-hostname: the value indicates the hostname of the endpoint. If the endpoint is derived from a k8s pod. This must match the hostname in pod spec.
+    // failure-domain.beta.kubernetes.io/zone: the value indicates the zone where the endpoint is located.
```

Comment:
_If_ we are codifying these, we should also add `topology.kubernetes.io/zone` and `topology.kubernetes.io/region` but would rather just copy the labels specified in `topologyKeys` in the service spec.

Reply:
xref https://github.com/kubernetes/enhancements/issues/837 for the new GA label keys for zone and region

Reply:
Cool. I missed that.